libraries/standard/json/cbmc/proofs/JSON_Search/JSON_Search_harness.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file JSON_Search_harness.c
+ * @brief Implements the proof harness for the JSON_Search function.
+ */
+
+#include <stdlib.h>
+#include "json.h"
+
+void harness()
+{
+    char * buf = NULL;
+    size_t max;
+    char * queryKey = NULL;
+    size_t queryKeyLength;
+    char separator;
+    char * outValue;
+    size_t outValueLength;
+    JSONStatus_t ret;
+
+    __CPROVER_assume( max < CBMC_MAX_BUFSIZE );
+
+    if( nondet_bool() )
+    {
+        buf = malloc( max );
+        __CPROVER_assume( __CPROVER_r_ok( buf, max ) );
+    }
+
+    __CPROVER_assume( queryKeyLength < CBMC_MAX_BUFSIZE );
+
+    if( nondet_bool() )
+    {
+        queryKey = malloc( queryKeyLength );
+        __CPROVER_assume( __CPROVER_r_ok( queryKey, queryKeyLength ) );
+    }
+
+    ret = JSON_Search( buf,
```

Comment:
We could assert that the value returned is one of the values within the enum return type - but, frankly, I thought that would already be proven.

I can make assertions about outValue and outValueLength in the success case.  Like this:
```
    if( ret == JSONSuccess )
    {
        __CPROVER_assert( ( outValue >= buf ) &&
                          ( ( outValue + outValueLength ) <= ( buf + max ) ),
                          "The output value is a sequence of characters within buf." );
    }
```
What do you think?

Reply:
Experimentation seems to show there is no builtin check on the enums.  I'll add them.

Reply:
There is a builtin check, but it had some false positives if the harness wasn't written carefully, we had to turn it off from the default set of flags.  You might want to try turning it on for your repo, since its all new proofs.
```
 --enum-range-check           checks that all enum type expressions have values in the enum range
```

Reply:
But having the explicit check is also good documentation, so I'd do that too.